### ISSUE TITLE
 Drastic emulator has no sound

### ISSUE THREAD

**Issue by Oberworld** [goweiwen/Allium#48]
At least on the miyoo mini plus, which is the one I have, the drastic emulator does not have sound.

I have also tried this other version that has a much friendlier menu interface and the result is the same:

[drastic-v1.6.zip](https://github.com/goweiwen/Allium/files/13256732/drastic-v1.6.zip)

(I recommend to update Allium's drastic to this one.)

Both versions of drastic work with sound in onion.

Thanks.

**Comment by pakrett** [comment]
I have a miyoo mini v2 and drastic has no sound either.

**Comment by XK9274** [comment]
Heya, this is due to mi_ao not receiving the notification to unmute and having its volume set to 0.  I had a similar hurdle with Pico8, but I believe steward handles mi_ao in the SDL libs where i handle them in the launch.sh

To fix this, add this function to your launch.sh:

It should be called just before the call for drastic to start. 

To explain, what this will do is launch, then sleep 3 seconds and wait for mi_ao to appear, when this character device/module appears it will unmute it and set a respectable volume limit. I don't know if @goweiwen handles mi_ao so i've hardcoded the volume limit. If this device doesn't appear after 30 seconds it will give up trying. 

```sh
set_snd_level() {
    sleep 3
    local target_vol="$1"
    local current_vol
    local start_time
    local elapsed_time

    start_time=$(date +%s)
    while [ ! -e /proc/mi_modules/mi_ao/mi_ao0 ]; do
        sleep 0.2
        elapsed_time=$(( $(date +%s) - start_time ))
        if [ "$elapsed_time" -ge 30 ]; then
            echo "Timed out waiting for /proc/mi_modules/mi_ao/mi_ao0" > /dev/pts/0 
            return 1
        fi
    done

    start_time=$(date +%s)
    while true; do
        echo "set_ao_mute 0" > /proc/mi_modules/mi_ao/mi_ao0
        echo "set_ao_volume 0 -9dB" > /proc/mi_modules/mi_ao/mi_ao0
        echo "set_ao_volume 1 -9dB" > /proc/mi_modules/mi_ao/mi_ao0
        current_vol=$(get_curvol)

        if [ "$current_vol" = "$target_vol" ]; then
            echo "Volume set to ${current_vol}dB" > /dev/pts/0
            return 0
        fi

        elapsed_time=$(( $(date +%s) - start_time ))
        if [ "$elapsed_time" -ge 360 ]; then
            echo "Timed out trying to set volume" > /dev/pts/0 
            return 1
        fi

        sleep 0.2
    done
}

set_snd_level &
./drastic
```

**Comment by hartleyshc** [comment]
Tried your fix @XK9274 

Works perfectly here on both DraStic 1.5 (bundled with Allium) and DraStic 1.6 (replaced 1.5 in the .allium folder). Thanks!

**Comment by XK9274** [comment]
> Tried your fix @XK9274 
> 
> Works perfectly here on both DraStic 1.5 (bundled with Allium) and DraStic 1.6 (replaced 1.5 in the .allium folder). Thanks! 

Great! 👍

**Comment by Oberworld** [comment]
Working here too. Thank you very much!! Great work!!

**Comment by XK9274** [comment]
Ah good to hear, glad it's working for everyone. 

If you had the longer version just FYI I've removed the second loop as it's not really required. Should work in it's edited form but let me know if not & we can add a sleep between the first loop ending and the echo's in 👍

**Comment by goweiwen** [comment]
Thanks for the solution, @XK9274!

I'll fix this in the next update along with updating to DraStic 1.6.

**Comment by Alessio81Pa** [comment]
> Heya, this is due to mi_ao not receiving the notification to unmute and having its volume set to 0. I had a similar hurdle with Pico8, but I believe steward handles mi_ao in the SDL libs where i handle them in the launch.sh
> 
> To fix this, add this function to your launch.sh:
> 
> It should be called just before the call for drastic to start.
> 
> To explain, what this will do is launch, then sleep 3 seconds and wait for mi_ao to appear, when this character device/module appears it will unmute it and set a respectable volume limit. I don't know if @goweiwen handles mi_ao so i've hardcoded the volume limit. If this device doesn't appear after 30 seconds it will give up trying.
> 
> ```shell
> set_snd_level() {
>     sleep 3
>     local start_time
>     local elapsed_time
> 
>     start_time=$(date +%s)
>     while [ ! -e /proc/mi_modules/mi_ao/mi_ao0 ]; do
>         sleep 0.2
>         elapsed_time=$(( $(date +%s) - start_time ))
>         if [ "$elapsed_time" -ge 30 ]; then
>             return 1
>         fi
>     done
> 
>     echo "set_ao_mute 0" > /proc/mi_modules/mi_ao/mi_ao0
>     echo "set_ao_volume 0 -9dB" > /proc/mi_modules/mi_ao/mi_ao0
>     echo "set_ao_volume 1 -9dB" > /proc/mi_modules/mi_ao/mi_ao0
> 
> }
> 
> set_snd_level &
> ./drastic
> ```
> 
> e* removed debug prints e2* removed currentvol logic, not required

where in the file should these lines be inserted?

**Comment by XK9274** [comment]
> > Heya, this is due to mi_ao not receiving the notification to unmute and having its volume set to 0. I had a similar hurdle with Pico8, but I believe steward handles mi_ao in the SDL libs where i handle them in the launch.sh
> > 
> > To fix this, add this function to your launch.sh:
> > 
> > It should be called just before the call for drastic to start.
> > 
> > To explain, what this will do is launch, then sleep 3 seconds and wait for mi_ao to appear, when this character device/module appears it will unmute it and set a respectable volume limit. I don't know if @goweiwen handles mi_ao so i've hardcoded the volume limit. If this device doesn't appear after 30 seconds it will give up trying.
> > 
> > ```shell
> > set_snd_level() {
> >     sleep 3
> >     local start_time
> >     local elapsed_time
> > 
> >     start_time=$(date +%s)
> >     while [ ! -e /proc/mi_modules/mi_ao/mi_ao0 ]; do
> >         sleep 0.2
> >         elapsed_time=$(( $(date +%s) - start_time ))
> >         if [ "$elapsed_time" -ge 30 ]; then
> >             return 1
> >         fi
> >     done
> > 
> >     echo "set_ao_mute 0" > /proc/mi_modules/mi_ao/mi_ao0
> >     echo "set_ao_volume 0 -9dB" > /proc/mi_modules/mi_ao/mi_ao0
> >     echo "set_ao_volume 1 -9dB" > /proc/mi_modules/mi_ao/mi_ao0
> > 
> > }
> > 
> > set_snd_level &
> > ./drastic
> > ```
> > 
> > e* removed debug prints e2* removed currentvol logic, not required
> 
> where in the file should these lines be inserted?

Insert the function set_snd_level anywhere aslong as it's not within the bounds of any other functions and it's defined before it's called. Then insert the call `set_snd_level &` just before the call to drastic as seen in the quoted example. 

Don't include the e* sections. 

I've been meaning to put together a version for weiwen that will automatically pull the correct volume level (this function hardcodes to -9db) as it's done in pico-8s launcher (which I'm happy for @goweiwen to use if he gets there before me)

**Comment by Alessio81Pa** [comment]
> > > Heya, this is due to mi_ao not receiving the notification to unmute and having its volume set to 0. I had a similar hurdle with Pico8, but I believe steward handles mi_ao in the SDL libs where i handle them in the launch.sh
> > > To fix this, add this function to your launch.sh:
> > > It should be called just before the call for drastic to start.
> > > To explain, what this will do is launch, then sleep 3 seconds and wait for mi_ao to appear, when this character device/module appears it will unmute it and set a respectable volume limit. I don't know if @goweiwen handles mi_ao so i've hardcoded the volume limit. If this device doesn't appear after 30 seconds it will give up trying.
> > > ```shell
> > > set_snd_level() {
> > >     sleep 3
> > >     local start_time
> > >     local elapsed_time
> > > 
> > >     start_time=$(date +%s)
> > >     while [ ! -e /proc/mi_modules/mi_ao/mi_ao0 ]; do
> > >         sleep 0.2
> > >         elapsed_time=$(( $(date +%s) - start_time ))
> > >         if [ "$elapsed_time" -ge 30 ]; then
> > >             return 1
> > >         fi
> > >     done
> > > 
> > >     echo "set_ao_mute 0" > /proc/mi_modules/mi_ao/mi_ao0
> > >     echo "set_ao_volume 0 -9dB" > /proc/mi_modules/mi_ao/mi_ao0
> > >     echo "set_ao_volume 1 -9dB" > /proc/mi_modules/mi_ao/mi_ao0
> > > 
> > > }
> > > 
> > > set_snd_level &
> > > ./drastic
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > e* removed debug prints e2* removed currentvol logic, not required
> > 
> > 
> > where in the file should these lines be inserted?
> 
> Insert the function set_snd_level anywhere aslong as it's not within the bounds of any other functions and it's defined before it's called. Then insert the call `set_snd_level &` just before the call to drastic as seen in the quoted example.
> 
> Don't include the e* sections.
> 
> I've been meaning to put together a version for weiwen that will automatically pull the correct volume level (this function hardcodes to -9db) as it's done in pico-8s launcher (which I'm happy for @goweiwen to use if he gets there before me)

don't work :(

**Comment by Alessio81Pa** [comment]
> > > Heya, this is due to mi_ao not receiving the notification to unmute and having its volume set to 0. I had a similar hurdle with Pico8, but I believe steward handles mi_ao in the SDL libs where i handle them in the launch.sh
> > > To fix this, add this function to your launch.sh:
> > > It should be called just before the call for drastic to start.
> > > To explain, what this will do is launch, then sleep 3 seconds and wait for mi_ao to appear, when this character device/module appears it will unmute it and set a respectable volume limit. I don't know if @goweiwen handles mi_ao so i've hardcoded the volume limit. If this device doesn't appear after 30 seconds it will give up trying.
> > > ```shell
> > > set_snd_level() {
> > >     sleep 3
> > >     local start_time
> > >     local elapsed_time
> > > 
> > >     start_time=$(date +%s)
> > >     while [ ! -e /proc/mi_modules/mi_ao/mi_ao0 ]; do
> > >         sleep 0.2
> > >         elapsed_time=$(( $(date +%s) - start_time ))
> > >         if [ "$elapsed_time" -ge 30 ]; then
> > >             return 1
> > >         fi
> > >     done
> > > 
> > >     echo "set_ao_mute 0" > /proc/mi_modules/mi_ao/mi_ao0
> > >     echo "set_ao_volume 0 -9dB" > /proc/mi_modules/mi_ao/mi_ao0
> > >     echo "set_ao_volume 1 -9dB" > /proc/mi_modules/mi_ao/mi_ao0
> > > 
> > > }
> > > 
> > > set_snd_level &
> > > ./drastic
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > e* removed debug prints e2* removed currentvol logic, not required
> > 
> > 
> > where in the file should these lines be inserted?
> 
> Insert the function set_snd_level anywhere aslong as it's not within the bounds of any other functions and it's defined before it's called. Then insert the call `set_snd_level &` just before the call to drastic as seen in the quoted example.
> 
> Don't include the e* sections.
> 
> I've been meaning to put together a version for weiwen that will automatically pull the correct volume level (this function hardcodes to -9db) as it's done in pico-8s launcher (which I'm happy for @goweiwen to use if he gets there before me)

[launch.txt](https://github.com/goweiwen/Allium/files/13709167/launch.txt)
this is my file

**Comment by goweiwen** [comment]
Thanks @XK9274 , this fix is included in v0.24.0 (bab48cc47775de34e26a837d88b867c6be2d68f8) along with DraStic 1.7.

**Comment by Alessio81Pa** [comment]
> Thanks @XK9274 , this fix is included in v0.24.0 ([bab48cc](https://github.com/goweiwen/Allium/commit/bab48cc47775de34e26a837d88b867c6be2d68f8)) along with DraStic 1.7.

But i've installed v1.7
